prow/resultstore/payload.go

```diff
@@ -0,0 +1,242 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resultstore
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/testgrid/metadata"
+	"google.golang.org/genproto/googleapis/devtools/resultstore/v2"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	v1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/kube"
+)
+
+type Payload struct {
+	Job       *v1.ProwJob
+	Started   *metadata.Started
+	Finished  *metadata.Finished
+	Files     []*resultstore.File
+	ProjectID string
+}
+
+func (p *Payload) invocation() (*resultstore.Invocation, error) {
+	if p.Job == nil {
+		return nil, errors.New("internal error: Payload.Job is nil")
+	}
+	i := &resultstore.Invocation{
+		Id:                   p.invocationID(),
+		StatusAttributes:     p.invocationStatusAttributes(),
+		Timing:               p.invocationTiming(),
+		InvocationAttributes: p.invocationAttributes(),
+		WorkspaceInfo:        p.workspaceInfo(),
+		Properties:           p.invocationProperties(),
+		Files:                p.Files,
+	}
+	return i, nil
+}
+
+func (p *Payload) invocationID() *resultstore.Invocation_Id {
+	return &resultstore.Invocation_Id{
+		// Name is a v4 UUID set in pjutil.go.
+		InvocationId: p.Job.Name,
+	}
+}
+
+func (p *Payload) invocationStatusAttributes() *resultstore.StatusAttributes {
+	status := resultstore.Status_TOOL_FAILED
+	switch p.Job.Status.State {
+	case v1.SuccessState:
+		status = resultstore.Status_PASSED
+	case v1.FailureState:
+		status = resultstore.Status_FAILED
+	case v1.AbortedState:
+		status = resultstore.Status_CANCELLED
+	case v1.ErrorState:
+		status = resultstore.Status_INCOMPLETE
+	}
+	return &resultstore.StatusAttributes{
+		Status: status,
+	}
+}
+
+func (p *Payload) invocationTiming() *resultstore.Timing {
+	start := p.Job.Status.StartTime.Time
+	duration := p.Job.Status.CompletionTime.Time.Sub(start)
+	return &resultstore.Timing{
+		StartTime: &timestamppb.Timestamp{
+			Seconds: start.Unix(),
+		},
+		Duration: &durationpb.Duration{
+			Seconds: int64(duration.Seconds()),
+		},
+	}
+}
+
+func (p *Payload) invocationAttributes() *resultstore.InvocationAttributes {
+	return &resultstore.InvocationAttributes{
+		// TODO: ProjectID might be assigned directly from the GCS
+		// BucketAttrs.ProjectNumber; requires a raw GCS client.
+		ProjectId:   p.ProjectID,
+		Labels:      []string{"prow"},
+		Description: descriptionFromLabels(p.Job.Labels),
+	}
+}
+
+func descriptionFromLabels(labels map[string]string) string {
+	jt := labels[kube.ProwJobTypeLabel]
+	repo := labels[kube.RepoLabel]
+	pull := labels[kube.PullLabel]
+	if ps := labels[kube.GerritPatchset]; ps != "" {
+		pull += "/" + ps
+	}
+	buildID := labels[kube.ProwBuildIDLabel]
+	job := labels[kube.ContextAnnotation]
+	return fmt.Sprintf("%s for %s/%s/%s/%s", jt, repo, pull, buildID, job)
+}
```

Comment:
nit: Should some of these become labels rather than just being in the description? I'm assuming labels are easier to search, filter, or group by.

Reply:
Many of these Prow labels become resultstore Properties. I believe ResultStore want their Labels to be low-radix since they are indexed. See https://github.com/googleapis/googleapis/blob/2aa3b1d5a90d05e0606d11787de475b0df068d49/google/devtools/resultstore/v2/invocation.proto#L166, so using any kind of ID is bad.

I intend in future work to allow configuration of custom resultstore labels by repo/job. Those are probably the only ones we should create.

Reply:
Many of these Prow labels become resultstore Properties. I believe ResultStore want their Labels to be low-radix since they are indexed. See https://github.com/googleapis/googleapis/blob/2aa3b1d5a90d05e0606d11787de475b0df068d49/google/devtools/resultstore/v2/invocation.proto#L166, so using any kind of ID is bad.

I intend in future work to allow configuration of custom resultstore labels by repo/job. Those are probably the only ones we should create.